js/src/shell/OSObject.cpp

```diff
@@ -322,6 +324,58 @@ static bool osfile_readRelativeToScript(JSContext* cx, unsigned argc,
   return ReadFile(cx, argc, vp, true);
 }
 
+static bool osfile_listDir(JSContext* cx, unsigned argc, Value* vp) {
+  CallArgs args = CallArgsFromVp(argc, vp);
+
+  if (args.length() != 1) {
+    JS_ReportErrorASCII(cx, "os.file.listDir requires 1 argument");
+    return false;
+  }
+
+  if (!args[0].isString()) {
+    JS_ReportErrorNumberASCII(cx, js::shell::my_GetErrorMessage, nullptr,
+                              JSSMSG_INVALID_ARGS, "os.file.listDir");
+    return false;
+  }
+
+  RootedString givenPath(cx, args[0].toString());
+  RootedString str(cx, ResolvePath(cx, givenPath, ScriptRelative));
+  if (!str) {
+    return false;
+  }
+
+  UniqueChars pathname = JS_EncodeStringToLatin1(cx, str);
+  if (!pathname) {
+    JS_ReportErrorASCII(cx, "os.file.listDir cannot convert path to Latin1");
+    return false;
+  }
+  DIR* dir = opendir(pathname.get());
+  if (!dir) {
+    JS_ReportErrorASCII(cx, "os.file.listDir is unable to open: %s",
+                        pathname.get());
+    return false;
+  }
+
+  RootedValueVector elems(cx);
+  while (struct dirent* entry = readdir(dir)) {
+    if (!(str = JS_NewStringCopyZ(cx, entry->d_name))) {
```

Comment:
This will return both files and directories in the same array.
that also includes "." and ".."
is this expected?  (I suppose the caller is responsible to handle/filter them?)

also, this doesn't work on Windows.
The code here could be copied.
https://searchfox.org/mozilla-central/rev/4df8821c1b824db5f40f381f48432f219d99ae36/js/src/jsapi-tests/testBinASTReader.cpp#190-214
https://searchfox.org/mozilla-central/rev/4df8821c1b824db5f40f381f48432f219d99ae36/js/src/jsapi-tests/testBinASTReader.cpp#449-453

Reply:
Yes, "." and ".." are included, and I think we can safely delegate this to the code using this interface.

I will look at your pointers, and modify this code accordingly.